### PR TITLE
Issue 15856 - Confusing error message with -transition=checkimports

### PR DESCRIPTION
**Pull request by 9rnsr**
Improve deprecation message mechanism for next minor release 2.071.1.

### CODE REVIEW

**Comment by WalterBright** [comment]
Do we really need to change 19 files to change a message? Something seems wrong.

**Comment by 9rnsr** [comment]
> Do we really need to change 19 files to change a message? Something seems wrong.

Yes we must fix the unsync between C++ headers and D declarations that was introduced in #5445. It's definitely wrong, because when a `search` function is used from both C++ and D code, the mismatch would cause different symbol search behaviors.

Today, `Scope.search` is actually used in `iasm.c`. Fortunately it has introduced no bugs, but the dangerous problem should be fixed ASAP.

File: test/fail_compilation/checkimports2c.d [hunk 1]
```diff
@@ -2,9 +2,11 @@
 /*
 TEST_OUTPUT:
 ---
-fail_compilation/checkimports2c.d(20): Deprecation: local import search method found variable imports.imp2.X instead of variable checkimports2c.X
-fail_compilation/checkimports2c.d(29): Deprecation: local import search method found variable imports.imp2.X instead of variable checkimports2c.X
-fail_compilation/checkimports2c.d(30): Deprecation: local import search method found variable imports.imp2.Y instead of variable imports.imp1.Y
+fail_compilation/checkimports2c.d(22): Deprecation: local import search method found variable imports.imp2.X instead of variable checkimports2c.X
+fail_compilation/checkimports2c.d(28): Deprecation: local import search method found variable imports.imp2.X instead of nothing
+fail_compilation/checkimports2c.d(29): Deprecation: local import search method found variable imports.imp2.Y instead of nothing
```
Comment (Geod24) [inline test/fail_compilation/checkimports2c.d]: Looking at the test case, those don't seem very clear.
Could we have something that isn't refering to the compiler internal, e.g:
`Deprecation: Variable imports.imp2.Y introduced by a private local import is not visible with the new import rule, use a public import or an alias` ?

Other suggestions welcome, but "found X instead of nothing" is not user friendly.

File: src/dmodule.d [hunk 2]
```diff
@@ -1093,10 +1093,13 @@ public:
             //        toChars(), ident->toChars(), flags, insearch, searchCacheSymbol ? searchCacheSymbol->toChars() : "null");
             return searchCacheSymbol;
         }
+
         uint errors = global.errors;
+
         insearch = 1;
         Dsymbol s = ScopeDsymbol.search(loc, ident, flags);
         insearch = 0;
+
```
Comment (Geod24) [inline src/dmodule.d]: Seems unrelated

File: src/dsymbol.d [hunk 3]
```diff
@@ -183,7 +183,7 @@ enum : int
                                     // meaning don't search imports in that scope,
                                     // because qualified module searches search
                                     // their imports
-    SearchCheck10378        = 0x40, // unqualified search with transition=checkimports switch
+    //SearchCheck10378        = 0x40, // unqualified search with transition=checkimports switch
```
Comment (Geod24) [inline src/dsymbol.d]: Since we use a version control system, what's the motivation to comment it instead of removing it ?

Comment (9rnsr) [inline src/dsymbol.d]: Because it was for the extra message printing around line 1312.

**Comment by WalterBright** [comment]
> Yes we must fix the unsync between C++ headers and D declarations that was introduced in #5445. It's definitely wrong, because when a  search  function is used from both C++ and D code, the mismatch would cause different symbol search behaviors.

This appears to be a separate issue from fixing a confusing error message. Hence, it should go in a separate PR.

Please, please, have only one issue per PR. Do not lump multiple things together. I've posted the reasons at least a dozen times. I don't think there's any risk of overflowing the PR # counter :-)

**Comment by 9rnsr** [comment]
@WalterBright Done. #5657

**Comment by 9rnsr** [comment]
When I've started to fix issue 15856, the default argument mismatch had blocked to change the default arguments to tweak search method in some places. Fortunately the final change didn't need that method, but it's still a serious bug, and related with the `search` code. That's why I put the commit in this PR.

File: src/mtype.d [hunk 4]
```diff
@@ -8639,7 +8663,31 @@ public:
             sc2.pop();
             return e;
         }
-        s = sym.search(e.loc, ident);
+
+        Dsymbol searchSym()
+        {
+            int flags = 0;
```
Comment (MartinNowak) [inline src/mtype.d]: No need to declare flags here, just use 0 and SearchLocalsOnly directly.

**Comment by MartinNowak** [comment]
This also seems to fix [Issue 15897 – private base class method not seen through derived class](https://issues.dlang.org/show_bug.cgi?id=15897) like my PR https://github.com/dlang/dmd/pull/5727, true?

File: src/mtype.d [hunk 5]
```diff
@@ -7868,14 +7868,38 @@ public:
                 // goto L1;
             }
         }
-        s = sym.search(e.loc, ident);
```
Comment (MartinNowak) [inline src/mtype.d]: Yes, I think we should get rid of the default flags for `Dsymbol.search`. They are even different for Dsymbol and ScopeDSymbol :o.

**Comment by MartinNowak** [comment]
> Do we really need to change 19 files to change a message? Something seems wrong.

It's changing 6 files, a few headers and tests. And it also adds the old vs. new import check to TypeClass and TypeStruct. Seems fairly reasonable to me.

File: src/dclass.d [hunk 6]
```diff
@@ -1120,7 +1120,17 @@ public:
...
+        auto s = ScopeDsymbol.search(loc, ident, flags);
+
+        if ((flags & (SearchLocalsOnly | SearchImportsOnly)) == 0)
+        {
+            // keep old way search
+        }
+        else
+        {
+            flags |= SearchLocalsOnly;
+        }
```
Comment (MartinNowak) [inline src/dclass.d]: Rather do
```d
if (!(flags & (SearchLocalsOnly | SearchImportsOnly))
    flags |= SearchLocalsOnly;
```

File: test/fail_compilation/checkimports2c.d (hunk 1 above)
Comment (MartinNowak) [inline test/fail_compilation/checkimports2c.d]: We don't know through which import we found the symbol, and it would be too much work to figure out.

**Comment by MartinNowak** [comment]
I've put my comments into a PR https://github.com/9rnsr/dmd/pull/6.

**Comment by 9rnsr** [comment]
I merged https://github.com/9rnsr/dmd/pull/6, and then fixed a breaking introduced by it.
